pkg/algorithm/nodeselect/v1alpha1/select_node.go

```diff
@@ -150,42 +115,94 @@ func (ac *Config) getCandidateNode(listBlockDevice *ndmapis.BlockDeviceList) (ma
+			nodeDisk := nodeBlockDeviceMap[nodeName]
 			// Add the current disk to the diskList for this node.
 			nodeDisk.Items = append(nodeDisk.Items, value.Name)
 		}
 	}
 	return nodeBlockDeviceMap, nil
 }
 
-func (ac *Config) selectNode(nodeBlockDeviceMap map[string]*blockDeviceList) *nodeBlockDevice {
+// selectNode returns node name and list of blockdevice(nodeBlockDevice) can be used to
+// provision cstor pools
+func (ac *Config) selectNode(nodeBlockDeviceMap map[string]*blockDeviceList) (*nodeBlockDevice, error) {
 	// selectedBlockDevice will hold a node capable to form pool and list of disks attached to it.
 	selectedBlockDevice := &nodeBlockDevice{
 		NodeName: "",
 		BlockDevices: blockDeviceList{
 			Items: []string{},
 		},
 	}
+	spcObj := spcv1alpha1.SPC{Object: ac.Spc}
 	// bdCount will hold the number of block devices that will be selected from a qualified
 	// node for specific pool type
 	var bdCount int
 	// minRequiredDiskCount will hold the required number of disk that should be selected from a qualified
 	// node for specific pool type
 	minRequiredDiskCount := spcv1alpha1.DefaultDiskCount[ac.poolType()]
+
+	// Below snippet is required to get the node and blockdevice topology from
+	// blockdeviceclaims which were created during previous reconciliation
+	if spcObj.IsAutoProvisioning() {
+		// List all blockdeviceclaims related to spc
+		// NOTE: This list contains blockdevices on which csps' are already
+		// created filter and use it
+		bdcList, err := bdc.NewKubeClient().
+			WithNamespace(ac.Namespace).
+			List(metav1.ListOptions{LabelSelector: string(apis.StoragePoolClaimCPK) + "=" + ac.Spc.Name})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to select node and blockdevices")
+		}
+		// customBDCList is a wrapper over blockdeviceclaim list
+		customBDCList := bdc.BlockDeviceClaimList{
+			ObjectList: bdcList,
+		}
+		// Form node and blockdevice topology for blockdevices which has bdc
+		claimedDevicesOnNode := customBDCList.GetBlockDeviceNamesByNode()
+		for claimedNodeName, bdNameList := range claimedDevicesOnNode {
+			qualifiedBDList := []string{}
+			filteredBlockDevices := nodeBlockDeviceMap[claimedNodeName]
+
+			//Check is this node is usable or not
+			if filteredBlockDevices != nil {
+				qualifiedBDList = util.ListIntersection(filteredBlockDevices.Items, bdNameList)
+				// check whether bdNameList contains partially created claimed BDs
+				// i.e BD count is less than required BDs
+				if len(qualifiedBDList) < minRequiredDiskCount {
+					count := minRequiredDiskCount - len(qualifiedBDList)
+					availableBDOnNode := util.ListDiff(filteredBlockDevices.Items, qualifiedBDList)
+					if len(availableBDOnNode) < count {
+						continue
+					}
+					qualifiedBDList = append(qualifiedBDList, availableBDOnNode[:count]...)
+				}
+				selectedBlockDevice.BlockDevices.Items = append(selectedBlockDevice.BlockDevices.Items, qualifiedBDList...)
```

Comment:
should we make this as
`selectedBlockDevice.BlockDevices.Items = append(selectedBlockDevice.BlockDevices.Items, qualifiedBDList[:minRequiredDiskCount]...)`

Reply:
Updated